packages/core/src/extensions/Blocks/api/block.ts

```diff
@@ -0,0 +1,259 @@
+import { DOMSerializer, Schema } from "prosemirror-model";
+import { Plugin } from "prosemirror-state";
+import { Attribute, Extension, Node } from "@tiptap/core";
+import { BlockNoteEditor } from "../../../BlockNoteEditor";
+import {
+  BlockConfig,
+  BlockSchema,
+  BlockSpec,
+  PropSchema,
+  TipTapNode,
+  TipTapNodeConfig,
+} from "./blockTypes";
+import styles from "../nodes/Block.module.css";
+
+function camelToDataKebab(str: string): string {
+  return "data-" + str.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
+}
+
+export function propsToAttributes<
+  BType extends string,
+  PSchema extends PropSchema,
+  ContainsInlineContent extends boolean,
+  BSchema extends BlockSchema
+>(
+  blockConfig: Omit<
+    BlockConfig<BType, PSchema, ContainsInlineContent, BSchema>,
+    "render"
+  >
+) {
+  const tiptapAttributes: Record<string, Attribute> = {};
+
+  Object.entries(blockConfig.propSchema).forEach(([name, spec]) => {
+    tiptapAttributes[name] = {
+      default: spec.default,
+      keepOnSplit: true,
+      // Props are displayed in kebab-case as HTML attributes. If a prop's
+      // value is the same as its default, we don't display an HTML
+      // attribute for it.
+      parseHTML: (element) => element.getAttribute(camelToDataKebab(name)),
+      renderHTML: (attributes) =>
+        attributes[name] !== spec.default
+          ? {
+              [camelToDataKebab(name)]: attributes[name],
+            }
+          : {},
+    };
+  });
+
+  return tiptapAttributes;
+}
+
+export function parse<
+  BType extends string,
+  PSchema extends PropSchema,
+  ContainsInlineContent extends boolean,
+  BSchema extends BlockSchema
+>(
+  blockConfig: Omit<
+    BlockConfig<BType, PSchema, ContainsInlineContent, BSchema>,
+    "render"
+  >
+) {
+  // TODO: This won't work for content copied outside BlockNote. Given the
+  //  variety of possible custom block types, a one-size-fits-all solution
+  //  probably won't work and we'll need an optional parseHTML option.
+  return blockConfig.parse
+    ? [
+        {
+          getAttrs: (node: HTMLElement | string) => {
+            if (typeof node === "string") {
+              return false;
+            }
+
+            return blockConfig.parse!(node);
+          },
+        },
+      ]
+    : [
+        {
+          tag: "div[data-content-type=" + blockConfig.type + "]",
+        },
+      ];
+}
+
+export function render<
```

Comment:
let's add a comment on what this function does and for which scenarios it does / doesn't work